[tengcharmaine] iP

tengcharmaine opened the pull request:
### BotChat

> "You can do anything, but not everything." - David Allen ([source](https://dansilvestre.com/productivity-quotes/))

BotChat frees your mind of having to remember things you need to do. It is
- text-based
- easy to learn
- ~~FAST~~ SUPERFAST to use

All you need to do is:
1. download it from [here](https://github.com/tengcharmaine/ip).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you :hand_over_mouth:

And it is **FREEEEEEEE**!!!!!

Features
- [x] Managing tasks
- [x] Managing different types of tasks (Event, Deadline, Todo)
- [x] Finding tasks
- [ ] Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here is the `main` method:
```ruby
public static void main(String[] args) {
        new BotChat(FILE_PATH).run();
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

tengcharmaine committed:
changed chatbot name + implemented skeletal version

tengcharmaine committed:
added echoing of user's input + exit program when bye is keyed in

tengcharmaine committed:
abstracted out the echo function into another method called userInput

tengcharmaine committed:
added addTask (add task to list) and listTasks (list all tasks) methods

tengcharmaine committed:
added the recommended Task.java class

tengcharmaine committed:
added markTask and unmarkTasks method

tengcharmaine committed:
added mark and unmark methods and overridden toString method

tengcharmaine committed:
created new classes for event, deadline and todo tasks

tengcharmaine committed:
added to the 3 newly created tasks types

tengcharmaine committed:
created method for creating event tasks

tengcharmaine committed:
updated addEventTask method to resolve bug

tengcharmaine committed:
updated addDeadlineTask method

tengcharmaine committed:
updated addTodoTask method

tengcharmaine committed:
updated methods to check for task description else throws error

tengcharmaine committed:
updated expected file to actual output

tengcharmaine committed:
added DukeException class for handling errors

tengcharmaine committed:
implemented a few error handling using DukeException

tengcharmaine committed:
added more error handling

tengcharmaine committed:
changed Task to ArrayList and added the deleteTask method

tengcharmaine committed:
editted handleinput method to use enums

tengcharmaine committed:
add file with todo list stored and update Duke with load and store into file methods

tengcharmaine committed:
edit loadTasksFromHardDisk() method to not print out files and add convertTask method to convert text to Task string format

tengcharmaine committed:
resolve bug that added spacing in description

tengcharmaine committed:
Update class to be able to read and convert date

tengcharmaine committed:
Fix bugs for the conversion of date

tengcharmaine committed:
Update chatbot to be able to read both date and time for deadline task

tengcharmaine committed:
Update Event task to convert date and time

tengcharmaine committed:
Create more classes for OOP + updated each class with the relevant methods

tengcharmaine committed:
Fix bugs in code (loading of tasks still shown though)

tengcharmaine committed:
Create botChat package

tengcharmaine committed:
Merge branch 'add-gradle-support' of https://github.com/tengcharmaine/ip

tengcharmaine committed:
Set up gradle to run botChat

tengcharmaine committed:
Created testcases for TaskList class

tengcharmaine committed:
Create testcases for Parser class

tengcharmaine committed:
Remove unecessary import from ParserTest

tengcharmaine committed:
Add javadocs for all classes and methods

tengcharmaine committed:
Update to adhere coding standards (botChat to BotChat)

tengcharmaine committed:
Update switch statements to fit coding standard

tengcharmaine committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/BotChat/BotChat.java
#	src/main/java/BotChat/Command.java
#	src/main/java/BotChat/Task.java
#	src/main/java/BotChat/TaskList.java
#	src/main/java/BotChat/Todo.java
#	src/main/java/botChat/botChatException.java
#	src/test/java/BotChat/TaskListTest.java

tengcharmaine committed:
Update BotChatException to have JavaDocs

tengcharmaine committed:
Update classes with methods for find command

tengcharmaine committed:
Implement checkstyle

tengcharmaine committed:
Create GUI for botchat

tengcharmaine committed:
Update methods to display text in GUI

tengcharmaine committed:
Update methods to show replies and error messages on GUI

tengcharmaine committed:
Update javadocs

ryanlimdx commented on src/main/java/BotChat/BotChat.java:
```diff
@@ -0,0 +1,59 @@
+        default:
+            return "";
+        }
+    }
```
Perhaps this section could be extracted into the parser class to handle inputs? Since in the parser class, it is described as "A utility class responsible for parsing user input and converting task information." Could be more apt in there!

ryanlimdx commented on src/main/java/BotChat/Storage.java:
I like how clean this class is! Good job :)

ryanlimdx commented on src/main/java/BotChat/TaskList.java:
```diff
@@ -0,0 +1,272 @@
+        }
+    }
+
+    /**
+     * Checks if the provided date string has a valid date format.
+     *
+     * @param by The date string to be checked.
+     * @return True if the date format is valid, false otherwise.
+     */
+    private boolean isValidDateFormat(String by) {
+        try {
+            LocalDate.parse(by, DateTimeFormatter.ofPattern("yyyy-MM-dd"));
+            return true;
+        } catch (DateTimeParseException e) {
+            try {
+                LocalDateTime.parse(by, DateTimeFormatter.ofPattern("yyyy-MM-dd HHmm"));
+                return true;
+            } catch (DateTimeParseException ex) {
+                return false;
+            }
+        }
+    }
```
Maybe a class for formatting date time could be helpful? Methods for error checking as well as parsing the date time string into DateTime formats etc could be stored in there.

ryanlimdx commented on src/main/java/BotChat/TaskList.java:
```diff
@@ -0,0 +1,272 @@
+    }
+
+    /**
+     * Adds a tod0 task to the task list based on the user input.
```
There may be a slight typo here where "todo" is typed as "tod0", it can be hard to distinguish sometimes 😰.

ryanlimdx commented on src/main/java/BotChat/Ui.java:
Perhaps a `line` variable could be used to store the line, "____________________________________________________________\n" to save yourself the trouble of keying in that many underscores? It would help with consistency as well!

ryanlimdx commented on src/main/java/BotChat/Event.java:
```diff
@@ -0,0 +1,81 @@
+     * @param by The date to be checked.
+     * @return True if the date is in a valid format, false otherwise.
+     */
+    private boolean isValidDateFormat(String by) {
+        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("MMM dd yyyy");
+        try {
+            LocalDate date = LocalDate.parse(by, formatter);
+            String formattedDateTime = date.format(formatter);
+            return formattedDateTime.equals(by);
+        } catch (Exception e) {
+            return false;
+        }
+    }
```
As an example for the review regarding a DateTime class in `TaskList`, these could be left for the Date time class to handle. It would help to clean up the code with less repetition. Well done overall!

ryanlimdx reviewed (commented):
Overall, I found your code easy to read for the most part. I noted some minor typos, and some classes are quite long. Consider splitting into smaller classes if that makes sense, such as having a DateTime class!

tengcharmaine committed:
Create a JAR file named BotChat.jar with 3rd party libraries bundled in

So that BotChat is able to work on computers that do not have JavaFX installed

Use gradle to bundle up the chatbot's codes and also the 3rd party libraries

tengcharmaine committed:
Add assertions to Task constructor method.

To help to document the assumption that tasks must have non-empty, non-null descriptions

To ensure that the input parameters for description is not-empty and non-null

The assertion makes sure that if it is violated during runtime, an assertion error will be thrown, providing valuable information for debugging

tengcharmaine committed:
Add assertions to methods in Storage class.

To help to document the assumption that tasks must have non-empty, non-null descriptions

To ensure that the filepath is non-null; file exists; tasklist non-null.

The assertion makes sure that if it is violated during runtime, an assertion error will be thrown, providing valuable information for debugging

tengcharmaine committed:
Add assertions to Event constructor and isValidDateFormat method.

To help to document the assumption that events must have non-empty, non-null descriptions and also non-empty non-null dates

To ensure that the descriptions are non-empty, non-null; dates also non-empty non-null

The assertion makes sure that if it is violated during runtime, an assertion error will be thrown, providing valuable information for debugging

tengcharmaine committed:
Add assertions to addTask and deleteTask methods. Additionally, added a private static final to remove magic constants.

To help to document the assumption that tasks to be added or deleted cannot be null/have invalid indexes.

To ensure that tasks are non-null and have valid index.

The assertion makes sure that if it is violated during runtime, an assertion error will be thrown, providing valuable information for debugging.

tengcharmaine committed:
Abstract out method to get the index of task.

To help improve the code quality especially the readability.

To ensure that each method is not too long and complicated.

Abstraction allows methods to be shorter by making methods call other methods.

tengcharmaine committed:
Made variables into constants and remove unused variables.

To help improve the code quality especially the readability.

To ensure that each method is more readable.

Changing variables into constants with meaningful names allows methods to be more readable.

tengcharmaine committed:
Made variables into constants.

To help improve the code quality especially the readability.

To ensure that each method is more readable.

Changing variables into constants with meaningful names allows methods to be more readable.

tengcharmaine committed:
Made variables into constants.

To help improve the code quality especially the readability.

To ensure that each method is more readable.

Changing variables into constants with meaningful names allows methods to be more readable.

tengcharmaine committed:
Merge pull request #1 from tengcharmaine/branch-A-CodeQuality

Merge branch-A-CodeQuality to master

tengcharmaine committed:
Add the help command.

To provide in-app guidance to users.

To ensure that users know the available commands and things that can be done.

tengcharmaine committed:
Merge pull request #2 from tengcharmaine/branch-C-Help

Merge branch-C-Help into master.

tengcharmaine committed:
Implement help command to display commands in a separate window

Added functionality to display the available commands in a separate window when the user enters the help command.

Modified the BotChat class to include a showHelpWindow method responsible for creating a new JavaFX stage to display the help message.

Enhances user experience by providing a clearer presentation of available commands.

tengcharmaine committed:
Add help command message.

Makes the interface more user friendly.

tengcharmaine committed:
Update help command message.

Makes the interface more user friendly.

tengcharmaine committed:
Update help command message.

Makes the text more readable.

tengcharmaine committed:
Edit size of popup window.

Makes the interface nicer.

tengcharmaine committed:
Update help window message to make the message more understandable.

tengcharmaine committed:
Update help command message to makes the text more accurate.

tengcharmaine committed:
Update welcome message to makes the text more user-friendly.

tengcharmaine committed:
Update help command message to makes the text more accurate.

tengcharmaine committed:
Removed unused statements and unnecessary comments to make the code more readable.

tengcharmaine committed:
Change the colour of the dialogue box to give the chatbot more personality

tengcharmaine committed:
Remove unnecessary java docs for getter method

tengcharmaine committed:
Update java docs of method to make it more readable

tengcharmaine committed:
Update the txt file to test the loading and storing of chatbot

tengcharmaine committed:
Update txt file to test loading and storing of botchat

tengcharmaine committed:
Update README.md

Update README documentation to include new features and improve readability

tengcharmaine committed:
Add a screenshot of Ui into docs directory

tengcharmaine committed:
Update location of main class and the name of the Jar file for ease of identification

tengcharmaine committed:
Update readme for BotChat for users to know how to set up and use the chatbot

tengcharmaine committed:
Merge pull request #4 from tengcharmaine/branch-readme

Update readme for BotChat for users to know how to set up and use the chatbot

tengcharmaine committed:
Update README.md to fix interface mistakes

tengcharmaine committed:
Update README.md on github page to match the website

tengcharmaine committed:
Refactor Javadoc comments in Launcher class for clarity

tengcharmaine committed:
Add JavaDoc for getTaskIndex method in TaskList to get task index

tengcharmaine committed:
Remove an unused method from Ui class in BotChat

tengcharmaine committed:
Add JavaDocs for start method in Main class of BotChat

tengcharmaine committed:
Update archiveFileName in shadowJar task of build gradle

tengcharmaine committed:
Fix parsing bugs in Parser class for converting task strings

tengcharmaine committed:
Update text file for BotChat with new information and tasks
